Developing workflow steps

tapphughesn opened the pull request:
Adds "workflow steps" which are specified in the -w switch. You can navigate between workflow steps with the space and backspace keys. ImageViewer also displays information about current workflow step in the bottom right-hand corner.

Example:

`ImageViewer -w p,Pluera,10,1,b,ROI1,#4B9CD3 input.mha`

This command launches ImageViewer with two workflow steps: one paint mode with label "Pleura", radius 10, color 1 (int), and a second step in Box mode with label name "ROI1", color #4B9CD3 (hex color code).

If you want to save your paint and box overlays on exit, you could use a command like:

`ImageViewer -w B,roi1,#FFFFFF,p,pleura,10,1,p,bone,10,2,b,roi2,#4B9CD3 -S prefix input.mha`

Possible issues/TODO:

- Drawn boxes and rulers are currently not visible while in paint mode
- Currently there is no way to delete a box or ruler once it is drawn (that I am aware of)
- Need to update the help information you get from pressing "h"
- Maybe we should standardize color format between the paint and box modes (right now they are int and hex color code)
- Boxes show area in mm<sup>3</sup> which doesn't make sense for 2D images

aylward commented:
Perhaps it is sufficient to specify a colormap index instead of a complete color for the box?   It seems inconsistent to have painting use a colormap index and boxes use a full color specification.  Additionally, a full color specification is rather tedious to give on a command line.   I like the idea of giving it color!  Just suggesting using the same colormap as painting.  Granted, you will need to copy the colormap definition to the box (it can't really be shared with paint), but it should be consistent 1=red, etc.

s

tapphughesn commented:
> Perhaps it is sufficient to specify a colormap index instead of a complete color for the box? It seems inconsistent to have painting use a colormap index and boxes use a full color specification. Additionally, a full color specification is rather tedious to give on a command line. I like the idea of giving it color! Just suggesting using the same colormap as painting. Granted, you will need to copy the colormap definition to the box (it can't really be shared with paint), but it should be consistent 1=red, etc.
> 
> s

@aylward 

Do you know where one could get that colormap? It must be somewhere in ITK, since the ImageViewer overlay is an itk::Image. The box color is a QColor, which right now we are constructing from a hex code. Thanks.

aylward commented:
https://github.com/KitwareMedical/ImageViewer/blob/master/QtImageViewer/QtGlSliceView.cxx#:~:text=cColorTable%20%3D%20ColorTableType%3A%3ANew,cColorTable-%3EUseDiscreteColors()%3B
and 
https://github.com/KitwareMedical/ImageViewer/blob/20f93f84c25639ac32dd5559314997997fa80efa/QtImageViewer/QtGlSliceView.cxx#L400

tapphughesn commented:
@aylward 

I just changed the color input for boxes to an int using the colormap. So a now a command looks like:

`ImageViewer -w p,Pluera,10,1,b,ROI1,2 -S prefix input.mha`

I also deleted some unnecessary print statements and comments.

aylward commented:
Nice!!!

On Fri, Jul 9, 2021 at 5:12 PM Nicholas Tapp-Hughes <
***@***.***> wrote:

> @aylward <https://github.com/aylward>
>
> I just changed the color input for boxes to an int using the colormap. So
> a now a command looks like:
>
> ImageViewer -w p,Pluera,10,1,b,ROI1,2 -S prefix input.mha
>
> I also deleted some unnecessary print statements and comments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KitwareMedical/ImageViewer/pull/99#issuecomment-877459119>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACEJL53BQGZXM5XATO3Z4LTW5Q5XANCNFSM5ABFAH3A>
> .
>


-- 
Stephen R. Aylward, Ph.D.
Senior Director of Strategic Initiatives
---
Kitware: *Delivering innovative, open source, scientific software.*

tapphughesn commented:
Just made a few obvious commits, including removing a "build" folder that was generated by vscode (whoops)
